### PR TITLE
controller: bml: Revive bml_set_wifi_credentials (part 2/2)

### LINKED ISSUE #1058: [TASK] Revive bml_set_wifi_credentials
## Description

`bml_set_wifi_credentials` is an old implementation for a CLI command for changing credentials for a specific radio. It needs to be revived to allow setting credentials for any agent in the network.

Currently, the CLI command is defined to accept the following parameters:
```
int bml_set_wifi_credentials(BML_CTX ctx, const char *ssid, const char *pass, int sec, int vap_id, int force)
```

The way to set credentials in EasyMesh is using autoconfig where the controller is configured per `alid`. Supported parameters are ssid, password (network key), encryption type, authentication type, operating class and whether to configure as bAP or fAP. For more information, refer to the WSC M2 message defined in the WSC specification.

Therefore, the `bml_set_wifi_credentials` command should be updated to support all the new parameters, and delete the unused ones (vap_id and force).
In addition, it should provide another option which is whether to send auto-config renew to all known agents so the new credentials are picked up.

The BML API should be refactored to allow setting credentials for a specific alid, with multiple operating classes.

## Detailed description

The BML cli should support the following operations:
* set credentials for alid
* clear credentials
* update credentials

### Examples

Set fronthaul 2.4G VAP
```
bml_set_wifi_credentials 11:22:33:44:55:66 multiap-24g maprocks1 8x
```
Set backhaul 5G VAP
```
bml_set_wifi_credentials 11:22:33:44:55:66 multiap-5g maprocks1 12x backhaul
```
Clear credentials for alid 11:22:33:44:55:66
```
bml_clear_wifi_credentials 11:22:33:44:55:66
```
Clear all credentials:
```
bml_clear_wifi_credentials
```
Trigger autoconfig renew to all alids:
```
bml_update_wifi_credentials
```
Trigger autoconfig renew to specific alid:
```
bml_update_wifi_credentials 11:22:33:44:55:66
```

### Proposed low-level API in son_master_thread
```
struct sWifiCredentials {
    std::string ssid;
    std::string network_key;
    int opclass;
    int auth_type;
    int enc_type;
    bool backhaul;
}

/**
 * @brief set wifi credentials for specific Alid
 * 
 * Sets wifi credentials for specific alid and sends autoconfig renew if requested.
 *
 * @param alid alid of the agent to configure
 * @param credentials credentials vector, can be empty for teardown.
 * @return true if successful, false otherwise 
 */
bool set_wifi_credentials(const sMacAddr &alid,
                                       std::vector<sWifiCredentials> &credentials);

/**
 * @brief trigger wifi credentials update
 * 
 * Triggers wifi credentials update using the controller's current configuration.
 * If alids is empty, autoconfig renew for all known alids is sent.
 * 
 * @param alids list of alids for which to trigger credentials update
 * @return true if successful, false otherwise 
 */
bool update_wifi_credentials(const std::vector<sMacAddr> &alids);
```

## Exit Criteria

Test all new BML commands on actual RAX40 device with a NUC controller.

### PR DESCRIPTION
**Pull request by VladyslavTupikin**
According to the set_credentials UML diagram need to do refactoring for `set_wifi_credentials` API and update command for `beerocks_cli`.
Replace `cmd_reply` in `test_ap_config_bss_tear_down_cli` on the call `bml_set_wifi_credentials` via beerocks_cli.

### CODE REVIEW

**Comment by VladyslavTupikin** [comment]
Please, take a look only on commits:

beerocks: tlvf: Modify cACTION_BML_WIFI_CREDENTIALS_SET_REQUEST message
controller: bml: Refactor set_wifi_credentials

**Comment by VladyslavTupikin** [comment]
Need to change target branch after merging PR par1,

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml [hunk 1]
```diff
@@ -217,7 +217,22 @@ cACTION_BML_CHANGE_MODULE_LOGGING_LEVEL_RESPONSE:
 
 cACTION_BML_WIFI_CREDENTIALS_SET_REQUEST:
   _type: class
-  params: sWifiCredentials
+  al_mac: sMacAddr
+  ssid:
+   _type: char 
+   _length: [ "beerocks::message::WIFI_SSID_MAX_LENGTH" ]
```
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml]: (nitpick) you could have used a dynamic length here, like for operating channels.:
```suggestion
  ssid_length:
    _type: uint8_t
    _length_var: True
  ssid:
   _type: char 
   _length: [ ssid_length ]
```

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml [hunk 2]
```diff
@@ -217,7 +217,22 @@ cACTION_BML_CHANGE_MODULE_LOGGING_LEVEL_RESPONSE:
...
+  network_key:
```
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml]: Same here.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml [hunk 3]
```diff
@@ -217,7 +217,22 @@ cACTION_BML_CHANGE_MODULE_LOGGING_LEVEL_RESPONSE:
...
+    _type: char 
+    _length: [ "beerocks::message::WIFI_PASS_MAX_LENGTH" ]
+  authentication_type: uint16_t #  eWscAuth
+  encryption_type: uint16_t #  eWscEncr
+  bss_type: uint8_t  #  eWscVendorExtSubelementBssType
+  operating_channel_size:
```
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml]: (nitpick)
```suggestion
  operating_classes_size:
```
(the name of the list field with _size appended)

File: controller/src/beerocks/bml/bml.cpp [hunk 4]
```diff
@@ -181,16 +182,54 @@ int bml_event_register_cb(BML_CTX ctx, BML_EVENT_CB cb)
     return pBML->register_event_cb(cb);
 }
 
-int bml_set_wifi_credentials(BML_CTX ctx, const char *ssid, const char *pass, int sec, int vap_id,
-                             int force)
+int bml_set_wifi_credentials(BML_CTX ctx, const char *al_mac, const char *ssid,
+                             const char *network_key, const char *operating_class,
```
Comment (arnout) [inline controller/src/beerocks/bml/bml.cpp]: (nitpick)
```suggestion
                             const char *network_key, const char *bands,
```
The string is not an operating class, it's a band specification.

File: controller/src/beerocks/bml/bml.cpp [hunk 5]
```diff
@@ -181,16 +182,54 @@ int bml_event_register_cb(BML_CTX ctx, BML_EVENT_CB cb)
...
+    if (network_key) {
+        wifi_credentials.network_key         = network_key;
+        wifi_credentials.authentication_type = WSC::eWscAuth::WSC_AUTH_WPA2PSK;
+        wifi_credentials.encryption_type     = WSC::eWscEncr::WSC_ENCR_AES;
+    } else {
+        wifi_credentials.authentication_type = WSC::eWscAuth::WSC_AUTH_OPEN;
+        wifi_credentials.encryption_type     = WSC::eWscEncr::WSC_ENCR_NONE;
+    }
+
+    if (operating_class) {
```
Comment (arnout) [inline controller/src/beerocks/bml/bml.cpp]: (nitpick) I would have written:
```
if (!operating_class)
    operating_class = "24g-5g";
wifi_credentials.operating_class = son::wireless_utils::string_to_wsc_oper_class(operating_class);
```

File: controller/src/beerocks/bml/bml.h [hunk 6]
```diff
@@ -140,16 +140,17 @@ int bml_event_register_cb(BML_CTX ctx, BML_EVENT_CB cb);
...
  *
  * @param [in] ctx BML Context.
+ * @param [in] al_mac The agent mac adress
  * @param [in] ssid[BML_NODE_SSID_LEN] The SSID for the network. ssid array length must be equal to BML_NODE_SSID_LEN
- * @param [in] pass[BML_NODE_PASS_LEN] The WPA2 passphrase for the network. pass array length must be equal to BML_NODE_PASS_LEN
- * @param [in] sec Wi-Fi security/encryption (BML_WLAN_SEC_*).
- * @param [in] vap_id ID of the requested VAP (0 for main VAP).
- * @param [in] force Force update even if not all registered IREs are online.
+ * @param [in] network_key[BML_NODE_PASS_LEN] The WPA2 passphrase for the network. pass array length must be equal to BML_NODE_PASS_LEN
+ * @param [in] operating_class The list of operating classes for 2.4G or 5G radio.
```
Comment (arnout) [inline controller/src/beerocks/bml/bml.h]: It's not actually a list of operating classes...

File: controller/src/beerocks/bml/bml.h [hunk 7]
```diff
@@ -140,16 +140,17 @@ int bml_event_register_cb(BML_CTX ctx, BML_EVENT_CB cb);
...
+ * @param [in] bss_type The type of base station subsystem (fronthaul, backhaul, fronthaul-backhaul)
```
Comment (arnout) [inline controller/src/beerocks/bml/bml.h]: Ahem. BSS stands for Basic Service Set. But nobody uses that, just say BSS. Or VAP, the two can be used pretty much interchangeably.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml [hunk 8]
```diff
@@ -217,7 +217,22 @@ cACTION_BML_CHANGE_MODULE_LOGGING_LEVEL_RESPONSE:
...
+    _type: uint8_t
+    _length_var: True
+  operating_channels:
```
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml]: It's operating classes, not operating channels.

File: controller/src/beerocks/bml/internal/bml_internal.h [hunk 9]
```diff
@@ -51,9 +53,22 @@ class bml_internal : public beerocks::socket_thread {
...
+    /**
+    * Updates the Wi-Fi credentials (for WPA2-Personal policy) for the 
+    * beerocks network.
+    *
+    * @param [in] ctx BML Context.
+    * @param [in] al_mac The agent mac adress
+    * @param [in] ssid[BML_NODE_SSID_LEN] The SSID for the network. ssid array length must be equal to BML_NODE_SSID_LEN
+    * @param [in] network_key[BML_NODE_PASS_LEN] The WPA2 passphrase for the network. pass array length must be equal to BML_NODE_PASS_LEN
+    * @param [in] operating_class The list of operating classes for 2.4G or 5G radio.
+    * @param [in] bss_type The type of base station subsystem (fronthaul, backhaul, fronthaul-backhaul)
```
Comment (arnout) [inline controller/src/beerocks/bml/internal/bml_internal.h]: This is not what is in the parameter list...

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp [hunk 10]
```diff
@@ -384,14 +384,16 @@ void cli_bml::setFunctionsMapAndArray()
...
-        " Security values: 0=No Wi-Fi Security, 1=WEP-64 Encryption, 2=WEP-128 Encryption, "
-        "3=WPA-Personal, 4=WPA2-Personal, 5=WPA-WPA2-Personal Encryption. Default value 4."
-        " Vap default vaule 0."
-        " [<1(True) or 0(False)>] for force. Default value 1",
+        "bml_set_wifi_credentials",
+        "<al_mac> <ssid> [<network_key>] [<operating_class>] [<bss_type>]",
+        "Sets WiFi credentials to the given AL-MAC "
+        "al_mac - agent mac address "
+        "ssid - service set identifier "
+        "network_key - password. Default empty value. "
```
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]:
```suggestion
        "network_key - password. If empty, configure open network. Default - empty. "
```

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp [hunk 11]
```diff
@@ -384,14 +384,16 @@ void cli_bml::setFunctionsMapAndArray()
...
+        "operating_class - can be 24g,5g or 24g-5g, list of operating classes. Default - 24g-5g. "
```
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]:
```suggestion
        "bands - can be 24g, 5g or 24g-5g. Default - 24g-5g. "
```

File: controller/src/beerocks/master/son_management.cpp [hunk 12]
```diff
@@ -1505,8 +1505,49 @@ void son_management::handle_bml_message(Socket *sd,
...
+
+        auto op_ch = request->operating_channels();
+        for (int i = 0; i < request->operating_channel_size(); i++) {
+            wifi_credentials.operating_class.push_back(uint16_t(op_ch[i]));
+        }
+
+        auto al_mac_addr = network_utils::mac_to_string(request->al_mac());
+        auto ires        = database.get_all_connected_ires();
+
+        if (ires.find(al_mac_addr) != ires.end()) {
```
Comment (arnout) [inline controller/src/beerocks/master/son_management.cpp]: In this case, the condition shouldn't be there. It should be possible to configure credentials for an agent before it starts up. Note that add_bss_info_configuration does not have such a check and doesn't add anything to the node - the bssinfo database is separate.

Actually, it was probably wrong to put the credentials in the database to begin with. It's configuration information, not topology information. But that's not for this PR.

File: tests/test_flows.py [hunk 13]
```diff
@@ -171,24 +171,22 @@ def test_ap_config_bss_tear_down_cli(self):
...
+        env.beerocks_cli_command('bml_set_wifi_credentials {} {} {} {} {}'.format(env.agents[0].mac,
+                                 "Multi-AP-24G-3-cli", "maprocks1", "24g", "fronthaul"))
         env.beerocks_cli_command('bml_update_wifi_credentials {}'.format(env.agents[0].mac))
 
         # Wait a bit for the renew to complete
         time.sleep(3)
 
         self.check_log(env.agents[0].radios[0],
-                       r"Received credentials for ssid: Multi-AP-24G-3 .* bss_type: 2")
+                       r"Received credentials for ssid: Multi-AP-24G-3-cli .* bss_type: 2")
```
Comment (arnout) [inline tests/test_flows.py:182]: Good catch!

**Review by arnout — APPROVED** [review]
All my comments are easy to fix, so approving.

Excellent commit message on the first commit!

**Review by tomereli — APPROVED** [review]
LGTM. Nicely done!

**Pull request merged**